From 1cc790ce376fc77a736e3430b32bf330bff13fb3 Mon Sep 17 00:00:00 2001
From: Michael Chang <mchang@suse.com>
Date: Mon, 8 Jun 2026 18:51:40 +0800
Subject: [PATCH 2/4] fs/btrfs: add btrfs-info command

Add new "btrfs-info" GRUB command to print basic filesystem metadata for
a given device, including label, UUID, total device count, and bytes
used.

Signed-off-by: Michael Chang <mchang@suse.com>
Reviewed-by: Neal Gompa <neal@gompa.dev>
Reviewed-by: Leo Sandoval <lsandova@redhat.com>
Part-of: <https://gitlab.freedesktop.org/gnu-grub/grub/-/merge_requests/160>
---
 grub-core/fs/btrfs.c | 103 ++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 92 insertions(+), 11 deletions(-)

diff --git a/grub-core/fs/btrfs.c b/grub-core/fs/btrfs.c
index 61dd1676a..5948530ff 100644
--- a/grub-core/fs/btrfs.c
+++ b/grub-core/fs/btrfs.c
@@ -41,6 +41,7 @@
 #include <grub/crypto.h>
 #include <grub/diskfilter.h>
 #include <grub/safemath.h>
+#include <grub/command.h>
 
 GRUB_MOD_LICENSE ("GPLv3+");
 
@@ -79,9 +80,11 @@ struct grub_btrfs_superblock
   grub_uint64_t generation;
   grub_uint64_t root_tree;
   grub_uint64_t chunk_tree;
-  grub_uint8_t dummy2[0x20];
+  grub_uint8_t dummy2[0x18];
+  grub_uint64_t bytes_used;
   grub_uint64_t root_dir_objectid;
-  grub_uint8_t dummy3[0x41];
+  grub_uint64_t num_devices;
+  grub_uint8_t dummy3[0x39];
   struct grub_btrfs_device this_device;
   char label[0x100];
   grub_uint8_t dummy4[0x100];
@@ -2282,6 +2285,20 @@ grub_btrfs_read (grub_file_t file, char *buf, grub_size_t len)
 				 data->tree, file->offset, buf, len);
 }
 
+static char *
+btrfs_unparse_uuid(struct grub_btrfs_data *data)
+{
+  return grub_xasprintf ("%04x%04x-%04x-%04x-%04x-%04x%04x%04x",
+			 grub_be_to_cpu16 (data->sblock.uuid[0]),
+			 grub_be_to_cpu16 (data->sblock.uuid[1]),
+			 grub_be_to_cpu16 (data->sblock.uuid[2]),
+			 grub_be_to_cpu16 (data->sblock.uuid[3]),
+			 grub_be_to_cpu16 (data->sblock.uuid[4]),
+			 grub_be_to_cpu16 (data->sblock.uuid[5]),
+			 grub_be_to_cpu16 (data->sblock.uuid[6]),
+			 grub_be_to_cpu16 (data->sblock.uuid[7]));
+}
+
 static grub_err_t
 grub_btrfs_uuid (grub_device_t device, char **uuid)
 {
@@ -2293,15 +2310,7 @@ grub_btrfs_uuid (grub_device_t device, char **uuid)
   if (!data)
     return grub_errno;
 
-  *uuid = grub_xasprintf ("%04x%04x-%04x-%04x-%04x-%04x%04x%04x",
-			  grub_be_to_cpu16 (data->sblock.uuid[0]),
-			  grub_be_to_cpu16 (data->sblock.uuid[1]),
-			  grub_be_to_cpu16 (data->sblock.uuid[2]),
-			  grub_be_to_cpu16 (data->sblock.uuid[3]),
-			  grub_be_to_cpu16 (data->sblock.uuid[4]),
-			  grub_be_to_cpu16 (data->sblock.uuid[5]),
-			  grub_be_to_cpu16 (data->sblock.uuid[6]),
-			  grub_be_to_cpu16 (data->sblock.uuid[7]));
+  *uuid = btrfs_unparse_uuid (data);
 
   grub_btrfs_unmount (data);
 
@@ -2431,6 +2440,71 @@ grub_btrfs_embed (grub_device_t device __attribute__ ((unused)),
 }
 #endif
 
+static grub_err_t
+grub_cmd_btrfs_info (grub_command_t cmd __attribute__ ((unused)), int argc,
+		     char **argv)
+{
+  grub_device_t dev;
+  char *devname;
+  struct grub_btrfs_data *data;
+  char *uuid;
+
+  if (argc < 1)
+    return grub_error (GRUB_ERR_BAD_ARGUMENT, "device name required");
+
+  devname = grub_file_get_device_name(argv[0]);
+
+  if (!devname)
+    return grub_errno;
+
+  dev = grub_device_open (devname);
+  grub_free (devname);
+  if (!dev)
+    return grub_errno;
+
+  data = grub_btrfs_mount (dev);
+  if (!data)
+    {
+      grub_device_close(dev);
+      return grub_error (GRUB_ERR_BAD_ARGUMENT, "failed to open fs");
+    }
+
+  if (*data->sblock.label != '\0')
+    {
+      char *label = grub_strndup (data->sblock.label, sizeof (data->sblock.label));
+      if (!label)
+	{
+	  grub_btrfs_unmount (data);
+	  grub_device_close (dev);
+	  return grub_errno;
+	}
+      grub_printf("Label: '%s' ", label);
+      grub_free (label);
+    }
+  else
+    grub_printf("Label: none ");
+
+  uuid = btrfs_unparse_uuid (data);
+  if (!uuid)
+    {
+      grub_btrfs_unmount (data);
+      grub_device_close (dev);
+      return grub_errno;
+    }
+
+  grub_printf (" uuid: %s\n\tTotal devices %" PRIuGRUB_UINT64_T
+	       " FS bytes used %" PRIuGRUB_UINT64_T "\n",
+	       uuid,
+	       grub_le_to_cpu64 (data->sblock.num_devices),
+	       grub_le_to_cpu64 (data->sblock.bytes_used));
+
+  grub_free (uuid);
+  grub_btrfs_unmount (data);
+  grub_device_close (dev);
+
+  return GRUB_ERR_NONE;
+}
+
 static struct grub_fs grub_btrfs_fs = {
   .name = "btrfs",
   .fs_dir = grub_btrfs_dir,
@@ -2446,13 +2520,20 @@ static struct grub_fs grub_btrfs_fs = {
 #endif
 };
 
+static grub_command_t cmd_info;
+
 GRUB_MOD_INIT (btrfs)
 {
   grub_btrfs_fs.mod = mod;
   grub_fs_register (&grub_btrfs_fs);
+  cmd_info = grub_register_command ("btrfs-info",
+				    grub_cmd_btrfs_info,
+				    "DEVICE",
+				    "Print BtrFS info about DEVICE.");
 }
 
 GRUB_MOD_FINI (btrfs)
 {
   grub_fs_unregister (&grub_btrfs_fs);
+  grub_unregister_command (cmd_info);
 }
-- 
2.55.0

